### ISSUE TITLE
Proxy semantic completions back to tabnine

### ISSUE THREAD

**Issue by rhino1998** [neoclide/coc-tabnine#25]
Allowing CoC to communicate semantic completions back to TabNine from other language servers would allow TabNine to learn better and provide better completions. For this to work, CoC would have to be able to act as a language server proxy of sorts.

**Comment by sQVe** [comment]
@rhino1998 Does any of the other plugins currently do this?

**Comment by rhino1998** [comment]
Afaik, none of the other TabNine clients do this. I'm also not sure how best to structure this. It might be possible to just forward TabNine's language client completion requests to a relevant language server, but I'm not sure how to handle the case of multiple language servers being active for one language type.

**Comment by sQVe** [comment]
@rhino1998 As far as I know, and from what I've read about the TabNine API, this might be impossible to implement. TabNine gives suggestions based on cursor position and the context around the cursor. There is currently no way to provide "meta" information, which if I'm understanding your feature request correctly, you would need to implement something like this.

**Comment by rhino1998** [comment]
@sQVe TabNine can also act as a Language Server client ([here](https://tabnine.com/blog/semantic/)). If semantic completion is enabled, it integrates results from language servers into its own results. I'm suggesting that this plugin should expose a language server interface that returns results provided by the rest of the configured CoC completion sources. TabNine would then query this plugin (which would be acting as a language server) to get additonal completion results and information. This would allow users to gain the benefits of TabNine's semantic completion feature without having multiple copies of the same language server running.

**Comment by sQVe** [comment]
@rhino1998 I feel that you're barking up the wrong tree. I suggest you raise this feature request at TabNine instead as they neef to implement this first. It would then be quite simple for us to implement this feature, especially given any example or more extensive API docs.

**Comment by rhino1998** [comment]
TabNine already has support for acting as a language client. In the TabNine configuration, you can configure language servers for TabNine to connect to. The required functionality on TabNine's side already exists. TabNine does expect the Language Server to be a child process that communicates over stdio, but it's pretty easy to make a language server proxy from stdio to TCP or a Unix socket.

**Comment by sQVe** [comment]
@rhino1998 Hmm. I think the bundling of external LSPs and making coc.nvim a LSP with the combined results are a big task. It would also have quite big performance implications.

There might very well be reasons for this not being done already. I still think this feature request should be raised at TabNine. Especially since the creator of TabNine also is the maintainer for a couple of the clients.

**Comment by chemzqm** [comment]
> CoC would have to be able to act as a language server proxy of sorts.

It can't for now, and I can't see any way TabNine could receive semantic completions from outside.

**Comment by rhino1998** [comment]
Yeah, I wrote a kinda hacky prototype today of the system I was suggesting. It required adding a way for extensions to listen for completion results. Then I added a basic TCP language server listener thing to coc-tabnine. The assumption was that TabNine would request from the proxy language server first, which would then respond to TabNine after the rests of the completions results resolved. This didn't work very well in practice. It adds latency to TabNine's completions and a lot of complexity to everything else for minimal gain.

 I got around the requirement for TabNine's language servers to be subprocesses by just doing `socat stdio tcp-connect:<addr>:<port>` to the language server proxy thing.
